src/IO/ReadBufferFromS3.cpp

```diff
@@ -42,16 +43,18 @@ ReadBufferFromS3::ReadBufferFromS3(
     UInt64 max_single_read_retries_,
     const ReadSettings & settings_,
     bool use_external_buffer_,
+    size_t offset_,
     size_t read_until_position_,
     bool restricted_seek_)
     : SeekableReadBufferWithSize(nullptr, 0)
     , client_ptr(std::move(client_ptr_))
     , bucket(bucket_)
     , key(key_)
     , max_single_read_retries(max_single_read_retries_)
+    , offset(offset_)
```

Comment:
btw wouldn't it be better to do just seek instead of passing offset to buffer? At least for me it seems better

Reply:
That's also valid, but I had two reasons for it:
1. It seems strange to be able to define right bound (read_until_position) and not to be able to do that with offset upfront.
2. Calling seek like that depends on the fact that the initialisation is lazy, something that is an implementation detail and not so obvious. If it weren't lazy, I would start downloading the file and then seek to it.

If you still think it makes sense to call seek I'll gladly do it.

Reply:
> It seems strange to be able to define right bound (read_until_position) and not to be able to do that with offset upfront.

I am also in doubt about that `read_until_position` is passed as an argument here and not with `setReadUntilPosition()` (both were added by me) because in almost all places it is actually done with `setReadUntilPosition()` so there is small inconsistency. I was planning to substitute this `read_until_position` with `setReadUntilPosition` in remaining places.

> Calling seek like that depends on the fact that the initialisation is lazy, something that is an implementation detail and not so obvious. If it weren't lazy, I would start downloading the file and then seek to it.

ah, right, I agree, let's leave it as it is then.